packages/response-ops/rule_form/src/hooks/use_rule_form_steps.tsx

```diff
@@ -0,0 +1,300 @@
+/*
+ * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
+ * or more contributor license agreements. Licensed under the "Elastic License
+ * 2.0", the "GNU Affero General Public License v3.0 only", and the "Server Side
+ * Public License v 1"; you may not use this file except in compliance with, at
+ * your election, the "Elastic License 2.0", the "GNU Affero General Public
+ * License v3.0 only", or the "Server Side Public License, v 1".
+ */
+
+import { omit } from 'lodash';
+import { EuiHorizontalRule, EuiSpacer, EuiStepsProps, EuiStepsHorizontalProps } from '@elastic/eui';
+import React, { useState, useMemo, useCallback } from 'react';
+import { useRuleFormState } from './use_rule_form_state';
+import { RuleActions } from '../rule_actions';
+import { RuleDefinition } from '../rule_definition';
+import { RuleDetails } from '../rule_details';
+import {
+  RULE_FORM_PAGE_RULE_ACTIONS_TITLE,
+  RULE_FORM_PAGE_RULE_DEFINITION_TITLE,
+  RULE_FORM_PAGE_RULE_DETAILS_TITLE,
+} from '../translations';
+import { hasActionsError, hasActionsParamsErrors, hasParamsErrors } from '../validation';
+import { RuleFormStepId } from '../constants';
+
+interface UseRuleFormStepsOptions {
+  /* Used to track steps that have been interacted with and should mark errors with 'danger' instead of 'incomplete' */
+  touchedSteps: Record<RuleFormStepId, boolean>;
+  /* Used to track the current step in horizontal steps, not used for vertical steps */
+  currentStep?: RuleFormStepId;
+}
+
+/**
+ * Define the order of the steps programmatically. Updating this array will update the order of the steps
+ * in all places needed.
+ */
+const STEP_ORDER = [RuleFormStepId.DEFINITION, RuleFormStepId.ACTIONS, RuleFormStepId.DETAILS];
```

Comment:
nit: Would we benefit from a `Map`?

Reply:
`Map`s aren't great at keeping track of what order things are in, I think an array is probably the move here